### PR TITLE
feat(bin, stages): metrics listener

### PR DESCRIPTION
**Pull request by shekhirin**
Foundation for https://github.com/paradigmxyz/reth/issues/2604. With `MetricsListener`, we will be able to report sync metrics from both pipeline and live syncs.

### CODE REVIEW

**Review by onbjerg — COMMENTED** [review]
since this is not using the existing pipeline events channel, I would note in the docs why that is/what the difference is

Also, this can potentially override the checkpoint metric, which means it will look like reth unwinds 1000s of blocks if it is stopped (when in fact it just never committed that progress) as discussed. We still need a metric that only marks the actual checkpoint.

**Review by onbjerg — APPROVED** [review]
Clarified out of band

File: crates/metrics/src/listener/mod.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+impl Future for MetricsListener {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        let Some(event) = ready!(this.events_rx.poll_recv(cx)) else {
+            return Poll::Ready(())
+        };
+
+        this.handle_event(event);
```
Comment (mattsse) [inline crates/metrics/src/listener/mod.rs:51]: this needs to be in a loop

the receiver must be pending before we can return pending

File: crates/metrics/src/listener/mod.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+pub mod events;
```
Comment (mattsse) [inline crates/metrics/src/listener/mod.rs:1]: move below imports

File: crates/metrics/src/listener/events/sync.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    StageCheckpoint {
+        /// Stage ID.
+        stage_id: StageId,
+        /// Stage checkpoint.
+        checkpoint: StageCheckpoint,
+        /// Maximum known block number reachable by this stage.
+        /// If specified, `entities_total` metric is updated.
+        max_block_number: Option<BlockNumber>,
+    },
+}
```
Comment (mattsse) [inline crates/metrics/src/listener/events/sync.rs:23]: I'm 100% convinced this belongs in the metrics crate.

I think of the metrics crate as somethings that provides support for metrics only.
imo this belongs somewhere else, perhaps stage-sync

File: crates/metrics/src/listener/mod.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+/// Collection of metrics.
+#[derive(Clone, Copy, Debug)]
+pub enum MetricEvent {
+    /// Pipeline and live sync.
+    Sync(SyncMetricEvent),
+}
+
+/// Metrics routine that listens to new metric events on the `events_rx` receiver.
+#[derive(Debug)]
+pub struct MetricsListener {
```
Comment (mattsse) [inline crates/metrics/src/listener/mod.rs:23]: what does this type do exactly?

**Review by mattsse — CHANGES_REQUESTED** [review]
I kinda get why we want this to be separate from the `PipelineEvent`, we just want optional extra info about checkpoints?

I'd prefer if we could move the new types out of the metrics crate, imo moving this to stages crate makes more sense

File: crates/metrics/src/listener/mod.rs (hunk 1 above)
Comment (shekhirin) [reply crates/metrics/src/listener/mod.rs]: right, fixed

File: crates/metrics/src/listener/mod.rs (hunk 4 above)
Comment (shekhirin) [reply crates/metrics/src/listener/mod.rs]: added

File: crates/metrics/src/listener/events/sync.rs (hunk 3 above)
Comment (shekhirin) [reply crates/metrics/src/listener/events/sync.rs]: fair, moved to `stages`. Live sync will use it later to send stage checkpoint updates.

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**